ffi/custom_passes.cpp

```diff
@@ -0,0 +1,902 @@
+                refop = NULL;
+            }
+        }
+        return mutated;
+    }
+
+    /**
+     * Find related decrefs to incref inside a basicblock in order
+     */
+    std::vector<CallInst*> findRelatedDecrefs(BasicBlock* bb, CallInst* incref) {
+        std::vector<CallInst*> res;
+        for (Instruction &ii : *bb) {
+        CallInst *call_inst;
+        if ((call_inst = isRelatedDecref(incref, &ii))){
+            res.push_back(call_inst);
+        } else {
+            continue;
+        }
+        }
+        return res;
+    }
+
+    CallInst* isRelatedDecref(CallInst *incref, Instruction *ii) {
+        // TODO: DRY
+        if (ii->getOpcode() == Instruction::Call) {
+            CallInst *call_inst = dyn_cast<CallInst>(ii);
+            Value *callee = call_inst->getCalledOperand();
+            if ( callee->getName() != "NRT_decref" ) {
+                return NULL;
+            }
+            if (incref->getArgOperand(0) != call_inst->getArgOperand(0)) {
+                return NULL;
+            }
+            return call_inst;
+        }
+        return NULL;
+    }
+
+    void getAnalysisUsage(AnalysisUsage &Info) const override {
+        Info.addRequired<DominatorTreeWrapperPass>();
+        Info.addRequired<PostDominatorTreeWrapperPass>();
+    }
+
+    bool isNonNullFirstArg(CallInst *call_inst){
+        auto val = call_inst->getArgOperand(0);
+        auto ptr = dyn_cast<ConstantPointerNull>(val);
+        return ptr == NULL;
+    }
+
+    bool hasAnyDecrefInNode(BasicBlock *bb) {
+
+        for (Instruction &ii: *bb) {
+            CallInst* refop = GetRefOpCall(&ii);
+            if (refop != NULL && IsDecRef(refop)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Pre-condition: head_node dominates tail_node
+     */
+    bool hasDecrefBetweenGraph(BasicBlock *head_node, BasicBlock *tail_node,
+                               SmallVector<BasicBlock*, 20> &stack) {
+        if (basicBlockInList(head_node, stack)) {
+            return false;
+        }
+        if (DEBUG_PRINT) {
+            errs() << "Check..." << head_node->getName() << "\n";
+        }
+
+        if (hasAnyDecrefInNode(head_node)) return true;
+
+        stack.push_back(head_node);
+        Instruction *term = head_node->getTerminator();
+        for (unsigned i=0; i < term->getNumSuccessors(); ++i) {
+            BasicBlock *child = term->getSuccessor(i);
+            if (child == tail_node)
+                continue;
+            // XXX: Recurse
+            if(hasDecrefBetweenGraph(child, tail_node, stack)){
+                return true;
+            }
+        }
+        return false;
+    }
+}; // end of struct RefPrunePass
+
+
+char RefNormalizePass::ID = 0;
+char RefPrunePass::ID = 0;
+
+size_t RefPrunePass::stats_per_bb = 0;
+size_t RefPrunePass::stats_diamond = 0;
+size_t RefPrunePass::stats_fanout = 0;
+size_t RefPrunePass::stats_fanout_raise = 0;
```

Comment:
NOTE: These are globals but are protected by llvmlite's LLVM threadlock around the context and also Numba's compiler lock. Worse case, there's a race condition on rwm and they under report.